src/cdk-experimental/scrolling/virtual-scroll-viewport.ts

```diff
@@ -83,13 +85,76 @@ export class CdkVirtualScrollViewport implements OnInit, OnDestroy {
   /** The size of the viewport (in pixels). */
   private _viewportSize = 0;
 
-  /** Whether this viewport is attached to a CdkVirtualForOf. */
-  private _isAttached = false;
+  /** The pending scroll offset to be applied during the next change detection cycle. */
+  private _pendingScrollOffset: number | null;
+
+  /** the currently attached CdkVirtualForOf. */
+  private _forOf: CdkVirtualForOf<any> | null;
 
   constructor(public elementRef: ElementRef, private _changeDetectorRef: ChangeDetectorRef,
-              private _ngZone: NgZone,
+              private _ngZone: NgZone, private _sanitizer: DomSanitizer,
               @Inject(VIRTUAL_SCROLL_STRATEGY) private _scrollStrategy: VirtualScrollStrategy) {}
 
+  ngOnInit() {
+    const viewportEl = this.elementRef.nativeElement;
+    Promise.resolve().then(() => {
```

Comment:
We do need to re-run change detection after we know the size of the viewport. It *might* be possible to avoid by just making up an initial viewport size (e.g. 500px), rendering as if it really was that, and then checking *outside* the zone what the real size is. If the real size turns out to be bigger we would still have to run change detection again.

I'm not sure its really worth the extra logic though... The performance cost of this shouldn't be too severe, it only happens once on component initialization. We also already have other components that do a similar thing on initialization.